registry/registry_pubkey_update.go

```diff
@@ -0,0 +1,359 @@
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/pelicanplatform/pelican/server_structs"
+)
+
+type RegisteredPrefixUpdate struct {
+	ClientNonce         string `json:"client_nonce"`
+	ClientPayload       string `json:"client_payload"`
+	ClientSignature     string `json:"client_signature"`
+	ClientPrevSignature string `json:"client_prev_signature"`
+
+	ServerNonce     string `json:"server_nonce"`
+	ServerPayload   string `json:"server_payload"`
+	ServerSignature string `json:"server_signature"`
+
+	Pubkey     json.RawMessage `json:"pubkey"`
+	PrevPubkey json.RawMessage `json:"prev_pubkey"`
+	AllPubkeys json.RawMessage `json:"all_pubkeys"`
+	Prefixes   []string        `json:"prefixes"`
+}
+
+func getKeyByKid(keySet jwk.Set, kidToFind string) (jwk.Key, error) {
+	for i := 0; i < keySet.Len(); i++ {
+		key, _ := keySet.Key(i)
+		kid, ok := key.Get("kid")
+		if ok && kid == kidToFind {
+			return key, nil
+		}
+	}
+	return nil, fmt.Errorf("key with kid %s not found in the set", kidToFind)
+}
+
+func constructUpdatedKeySet(keySet jwk.Set, kidToRemove string, keyToAdd jwk.Key) error {
+	keyToRemove, err := getKeyByKid(keySet, kidToRemove)
+	if err != nil {
+		log.Warnf("Key %s not found in the key set, skipping removal: %v", kidToRemove, err)
+	} else {
+		// Remove the key only if it exists
+		if err := keySet.RemoveKey(keyToRemove); err != nil {
+			return errors.Wrap(err, "failed to remove key from set")
+		}
+	}
+	if err := keySet.AddKey(keyToAdd); err != nil {
+		return errors.Wrap(err, "failed to add key to the key set")
+	}
+	return nil
+}
+
+// Generate server nonce for key-sign challenge when updating the public key of registered namespace(s)
+func updateNsKeySignChallengeInit(data *RegisteredPrefixUpdate) (map[string]interface{}, error) {
+	serverNonce, err := generateNonce()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to generate nonce for key-sign challenge")
+	}
+
+	serverPayload := []byte(data.ClientNonce + data.ServerNonce)
+
+	privateKey, err := loadServerKeys()
+	if err != nil {
+		return nil, errors.Wrap(err, "Server is unable to generate a key sign challenge: Failed to load the server's private key")
+	}
+
+	serverSignature, err := signPayload(serverPayload, privateKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to sign payload for key-sign challenge")
+	}
+
+	res := map[string]interface{}{
+		"server_nonce":     serverNonce,
+		"client_nonce":     data.ClientNonce,
+		"server_payload":   hex.EncodeToString(serverPayload),
+		"server_signature": hex.EncodeToString(serverSignature),
+	}
+	return res, nil
+}
+
+// Update the public key of registered prefix(es) if the http request passed client and server verification for nonce.
+// It returns the response data, and an error if any
+func updateNsKeySignChallengeCommit(ctx *gin.Context, data *RegisteredPrefixUpdate) (map[string]interface{}, error) {
```

Comment:
I think the design here is still a bit confused.  For example, it _removes_ the previous key -- not clear that is matching the intent here.

Let me suggest something cleaner:

1.  At the start of handshake, the registry sends all known public keys for the namespace.
2. The client receives the list of public keys and selects a corresponding private key from that list (if one is available) to sign the challenge.
3. The client sends the server the updated set of public keys plus the proof of possession from one of the known keys.

Thus, at the end of the handshake:
- The client has demonstrated it owns a private key from the existing known public keys.
- The client's desired set of keys is synchronized with the registry; the client can remove or add any arbitrary keys in the call.